### PR TITLE
NIFI-8522 NiFi can duplicate controller services when generating temp…

### PR DESCRIPTION
**Pull request by timeabarna**
Fix controller service duplication when generating templates
https://issues.apache.org/jira/browse/NIFI-8522

#### Description of PR

Fixing controller service duplication

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/util/SnippetUtils.java [hunk 1]
```diff
@@ -235,23 +234,41 @@ public FlowSnippetDTO populateFlowSnippet(final Snippet snippet, final boolean r
...
+        final ProcessGroupDTO highestProcessGroupDTO = dtoFactory.createProcessGroupDto(processGroup, recurse);
+        final Set<ProcessGroupDTO> processGroups = highestProcessGroupDTO.getContents().getProcessGroups();
+        fillContentsByGroupMap(highestProcessGroupDTO, contentsByGroup);
+
+        // maintain a listing of visited groups starting with each group in the snippet. this is used to determine
+        // whether a referenced controller service should be included in the resulting snippet. if the service is
+        // defined at groupId or one of it's ancestors, its considered outside of this snippet and will only be included
+        // when the includeControllerServices is set to true. this happens above when considering the processors in this snippet
+        final Set<String> visitedGroupIds = new HashSet<>();
+        for (String groupIdentifier : contentsByGroup.keySet()) {
```
Comment (gresockj) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/util/SnippetUtils.java:246]: Let's declare groupIdentifier as final

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/util/SnippetUtils.java [hunk 2]
```diff
@@ -299,62 +315,10 @@ public FlowSnippetDTO populateFlowSnippet(final Snippet snippet, final boolean r
...
-            .collect(Collectors.toMap(ComponentDTO::getId, childGroupDto -> childGroupDto));
-
-        for (final ProcessGroup childGroup : group.getProcessGroups()) {
-            final ProcessGroupDTO childDto = childGroupMap.get(childGroup.getIdentifier());
-            if (childDto == null) {
-                continue;
-            }
-
-            addControllerServices(childGroup, childDto, allServicesReferenced, includeControllerServices, visitedGroupIds, contentsByGroup, highestGroupId);
+    private void fillContentsByGroupMap(ProcessGroupDTO processGroup, Map<String, FlowSnippetDTO> contentByGroupMap) {
```
Comment (gresockj) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/util/SnippetUtils.java:318]: We can declare these parameters as final

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/util/SnippetUtils.java [hunk 3]
```diff
@@ -299,62 +315,10 @@ public FlowSnippetDTO populateFlowSnippet(final Snippet snippet, final boolean r
...
+        for (ProcessGroupDTO group: processGroup.getContents().getProcessGroups()) {
```
Comment (gresockj) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/util/SnippetUtils.java:319]: final group

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/util/SnippetUtils.java [hunk 4]
```diff
@@ -235,23 +234,41 @@ public FlowSnippetDTO populateFlowSnippet(final Snippet snippet, final boolean r
...
-                // whether a referenced controller service should be included in the resulting snippet. if the service is
-                // defined at groupId or one of it's ancestors, its considered outside of this snippet and will only be included
-                // when the includeControllerServices is set to true. this happens above when considering the processors in this snippet
-                final Set<String> visitedGroupIds = new HashSet<>();
-                addControllerServices(childGroup, childGroupDto, allServicesReferenced, includeControllerServices, visitedGroupIds, contentsByGroup, processGroup.getIdentifier());
+        final ProcessGroupDTO highestProcessGroupDTO = dtoFactory.createProcessGroupDto(processGroup, recurse);
+        final Set<ProcessGroupDTO> processGroups = highestProcessGroupDTO.getContents().getProcessGroups();
+        fillContentsByGroupMap(highestProcessGroupDTO, contentsByGroup);
+
+        // maintain a listing of visited groups starting with each group in the snippet. this is used to determine
```
Comment (gresockj) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/util/SnippetUtils.java:241]: For style purposes, can we capitalize each sentence?

**Review by gresockj — CHANGES_REQUESTED** [review]
Thanks @timeabarna for this contribution!  I was able to reproduce the issue, and observed that this PR does indeed resolve it.  I think the resulting code looks a bit clearer than the previous code in this section, as well.  Only some very minor style suggestions.

**Comment by timeabarna** [comment]
@gresockj thanks for your help, I've modified the code based on your recommendations.

**Review by gresockj — APPROVED** [review]

**Review by pvillard31 — APPROVED** [review]

**Comment by pvillard31** [comment]
Thanks @timeabarna @gresockj !

**Pull request closed without merging**